selftest/selftest.py

```diff
@@ -854,19 +879,160 @@ def parse_azure_nvme_id_version(azure_nvme_id_version_output: str) -> str:
         return parts[1]
 
 
+@dataclass
+class NetworkInterface:
+    """Network interface."""
+
+    name: str
+    driver: str
+    mac: str
+    ipv4_addrs: List[str]
+    udev_properties: Dict[str, str]
+
+
+@dataclass(eq=True, repr=True)
+class NetworkInfo:
+    """Network information."""
+
+    interfaces: Dict[str, NetworkInterface] = field(default_factory=dict)
+
+    @classmethod
+    def enumerate_interfaces(cls) -> Dict[str, NetworkInterface]:
+        """Retrieve all Ethernet interfaces on the system."""
+        interfaces: Dict[str, NetworkInterface] = {}
+        interface_names = [
+            interface
+            for interface in os.listdir(SYS_CLASS_NET)
+            if os.path.exists(os.path.join(SYS_CLASS_NET, interface, "device"))
+        ]
+
+        for interface_name in interface_names:
+            sys_path = Path(SYS_CLASS_NET, interface_name)
+            udev_properties = cls.query_udev_properties(interface_name)
+            driver_path = Path(sys_path, "device", "driver")
+            if not driver_path.is_symlink():
+                logger.debug(
+                    "ignoring interface %s without driver symlink", interface_name
+                )
+                continue
+
+            link = os.readlink(driver_path)
+            driver = os.path.basename(link)
+            mac = (sys_path / "address").read_text().strip()
+            ipv4_addrs = cls.get_ipv4_addresses(interface_name)
+            interfaces[interface_name] = NetworkInterface(
+                name=interface_name,
+                driver=driver,
+                mac=mac,
+                ipv4_addrs=ipv4_addrs,
+                udev_properties=udev_properties,
+            )
+
+        return interfaces
+
+    @staticmethod
+    def get_ipv4_addresses(interface: str) -> List[str]:
```

Comment:
nit: since you have defined NetworkInterface and there are other functions that use interface as type NetworkInterface, I would recommend using `interface_name` as argument to make it easier to distinguish between the two